### DISCUSSION TITLE
Converting a custom model into Pybamm class logic

### DISCUSSION THREAD

**Discussion by PaulKranefeld** [pybamm-team/PyBaMM#2603] [Q&A] [answered]
Hello, 
can someone describe me the basic procedure to convert a user defined model into a class structure that is compatible with Pybamm? This should make Pybamm functionalities like the experiment class accessible. I don't need a step by step guide but just a general approach and I am aware that this will be a time consuming task. 
Many thanks in advance.

**Comment by YannickNoelStephanKuhn** [comment, marked as answer]
Context: pybamm.Experiment handles the current control with a modular submodel, hence why "just" making a class like https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/full_battery_models/lithium_ion/basic_dfn.py is not enough.

You want to use the structure of https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/submodels/base_submodel.py for your class; I will call it MyClass.

Then, you need to make a second class following https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/full_battery_models/lithium_ion/base_lithium_ion_model.py.

The line that makes it compatible with pybamm.Experiment is https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/models/full_battery_models/lithium_ion/base_lithium_ion_model.py#L69, `self.set_external_circuit_submodel()`, as that registers the current control model. You won't need any of the other setters, instead use:
```self.submodels["my_equations"] = MyClass(self.param)```

**Reply by PaulKranefeld** [reply]
Thank you for the answer, this gives me a clear direction in which to continue my work.

**Reply by valentinsulzer** [reply]
Thanks @YannickNoelStephanKuhn . @PaulKranefeld , this is probably a common question so it would be great if you could document how you end up doing this (e.g. in a jupyter notebook)

**Reply by rtimms** [reply]
It is a little out of date (last updated August 2021), but there is a branch where we started some Li-S stuff using a similar approach, see [here](https://github.com/pybamm-team/PyBaMM/tree/base-lithium-sulfur/pybamm/models/full_battery_models/lithium_sulfur).

**Reply by PaulKranefeld** [reply]
Thank you very much for the tips. I will try my hand at it and post my results here if I am successful.

**Reply by Mrzhang-hub** [reply]
I wanna ask if you have  done this? I really want to know the detailed procedure how to create a custom model.
